### PR TITLE
check_ha use ComponentState

### PR DESCRIPTION
**Pull request by utopiabound**
check_ha use ComponentState
    
convert current check_ha to get_ha_resource_list()

Issue #1190 #1232

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/1288)

### CODE REVIEW

File: iml-agent/src/action_plugins/check_ha.rs [hunk 1]
```diff
@@ -71,27 +92,208 @@ impl AgentInfo {
...
+
+async fn check_corosync() -> Result<ComponentState<bool>, ImlAgentError> {
+    let mut corosync = ComponentState::<bool> {
+        ..Default::default()
+    };
+
+    if file_exists("/etc/corosync/corosync.conf").await {
+        corosync.config = ConfigState::Unknown;
+        corosync.service = systemd_unit_servicestate("corosync").await?;
+        // @@ check corosync setup
```
Comment (johnsonw) [inline iml-agent/src/action_plugins/check_ha.rs]: Is this a "Todo" and if so can it be done in a separate patch or should it be done in this one?

File: iml-agent/src/action_plugins/check_ha.rs [hunk 2]
```diff
@@ -71,27 +92,208 @@ impl AgentInfo {
...
+}
+
+/// property: org.freedesktop.systemd1
+/// format!("/org/freedesktop/systemd1/unit/{}_2eservice", name)
+/// UnitFileState   +   ActiveState  -> RunState
+/// disabled            inactive        Stopped
+/// enabled             inactive        Enabled
+/// disabled            active          Started
+/// enabled             active          Setup
+async fn systemd_unit_props(name: &str) -> Result<RunState, ImlAgentError> {
```
Comment (jgrund) [inline iml-agent/src/action_plugins/check_ha.rs]: Instead of unwrapping, use the ? operator to return early:

```rust
async fn systemd_unit_props(name: &str) -> Result<RunState, ImlAgentError> {
    let s = format!("/org/freedesktop/systemd1/unit/{}_2eservice", name);
    let output = cmd_output(
        "busctl",
        &[
            "--system",
            "--no-pager",
            "get-property",
            "org.freedesktop.systemd1",
            &s,
            "org.freedesktop.systemd1.Unit",
            "UnitFileState",
            "ActiveState",
        ],
    )
    .compat()
    .await?;

    let res: Vec<&str> = str::from_utf8(&output.stdout)?
        .lines()
        .map(|o| busctl_to_string(o))
        .collect::<Result<Vec<_>, ImlAgentError>>()?;

    match res[..] {
        ["disabled", "inactive"] => Ok(RunState::Stopped),
        ["enabled", "inactive"] => Ok(RunState::Enabled),
        ["disabled", "active"] => Ok(RunState::Started),
        ["enabled", "active"] => Ok(RunState::Setup),
        _ => Err(ImlAgentError::from(RequiredError(format!(
            "Unknown busctl ({}) output: {:?}",
            s, output.stdout
        )))),
    }
}
```

File: iml-agent/src/action_plugins/check_ha.rs [hunk 3]
```diff
@@ -71,27 +92,208 @@ impl AgentInfo {
...
+
+async fn systemd_unit_servicestate(name: &str) -> Result<ServiceState, ImlAgentError> {
+    if systemd_unit_enabled(name).await? {
+        Ok(ServiceState::Configured(systemd_unit_props(name).await?))
+    } else {
+        Ok(ServiceState::Unconfigured)
+    }
+}
+
+async fn file_exists(path: &str) -> bool {
```
Comment (jgrund) [inline iml-agent/src/action_plugins/check_ha.rs]: Why not use: https://docs.rs/tokio/0.1.22/tokio/fs/fn.metadata.html

File: iml-agent/src/action_plugins/check_ha.rs [hunk 4]
```diff
@@ -71,27 +92,208 @@ impl AgentInfo {
...
+    match cmd_output("test", &["-f", path]).compat().await {
+        Err(_) => false,
+        Ok(o) => o.status.success(),
+    }
+}
+
+/// systemctl is-enabled SERVICE :: call: org.freedesktop.systemd1
```
Comment (jgrund) [inline iml-agent/src/action_plugins/check_ha.rs]: Should this comment be here?

File: iml-agent/src/action_plugins/check_ha.rs [hunk 5]
```diff
@@ -71,27 +92,208 @@ impl AgentInfo {
...
+        Ok(ServiceState::Unconfigured)
+    }
+}
+
+async fn file_exists(path: &str) -> bool {
+    match cmd_output("test", &["-f", path]).compat().await {
+        Err(_) => false,
+        Ok(o) => o.status.success(),
+    }
+}
```
Comment (jgrund) [inline iml-agent/src/action_plugins/check_ha.rs]: These systemd functions should be in systemd.rs.

File: iml-agent/src/action_plugins/check_ha.rs (hunk 3 above)
Comment (utopiabound) [reply iml-agent/src/action_plugins/check_ha.rs]: I got errors, something about running in tokio threads. (I don't remember exactly)

Comment (jgrund) [reply iml-agent/src/action_plugins/check_ha.rs]: ah, right, needs to be inside a runtime. Let's wait till 0.2

File: iml-agent/src/action_plugins/check_ha.rs (hunk 1 above)
Comment (jgrund) [reply iml-agent/src/action_plugins/check_ha.rs]: Looks like this is still a todo?

Comment (johnsonw) [reply iml-agent/src/action_plugins/check_ha.rs]: Just following up on this. I'm assuming there is still some checks that need to be added?

**Comment by johnsonw** [comment]
Still wondering if there are some checks that need to be added based on the comment.

File: iml-agent/src/action_plugins/check_ha.rs [hunk 6]
```diff
@@ -2,94 +2,234 @@
...
+
+fn process_resource_list(output: &[u8]) -> Result<Vec<ResourceAgentInfo>, ImlAgentError> {
+    let element = Element::from_reader(output)?;
+
+    Ok(element
+        .find_all("group")
+        .flat_map(|g| {
+            let name = g.get_attr("id").unwrap_or("").to_string();
```
Comment (jgrund) [inline iml-agent/src/action_plugins/check_ha.rs]: Is `id` a required param? Is there a valid case for it to be missing?

File: iml-agent/src/action_plugins/check_ha.rs [hunk 7]
```diff
@@ -2,94 +2,234 @@
...
-    pub standard: String,         // e.g. ocf, lsb, stonith, etc..
-    pub provider: Option<String>, // e.g. heartbeat, lustre, chroma
-    pub ocftype: String,          // e.g. Lustre, ZFS
+use iml_wire_types::{
+    ComponentState, ConfigState, ResourceAgentInfo, ResourceAgentType, ServiceState,
+};
+use std::collections::HashMap;
+use tokio::fs::metadata;
+
+fn create(elem: &Element, group: Option<String>) -> ResourceAgentInfo {
```
Comment (jgrund) [inline iml-agent/src/action_plugins/check_ha.rs]: This could be:

`fn create<'a>(elem: &Element, group: impl Into<Option<&'a str>>) -> ResourceAgentInfo`

which would simplify callers.

File: iml-agent/src/action_plugins/check_ha.rs (hunk 6 above)
Comment (utopiabound) [reply iml-agent/src/action_plugins/check_ha.rs]: `id` is a required parameter.

File: iml-agent/src/action_plugins/check_ha.rs (hunk 7 above)
Comment (utopiabound) [reply iml-agent/src/action_plugins/check_ha.rs]: Or it's okay as it is...  The callers are not that complicated.

Comment (jgrund) [reply iml-agent/src/action_plugins/check_ha.rs]: Yeah, it works fine without it. I think it makes more sense in the big picture to push the complexity into trait bounds as the code either needs to repeat the same patterns each time or have them handled by the bound.
